### PR TITLE
Added support for http adapter config

### PR DESCRIPTION
**Pull request by venu1412**
### Checklist
- [x] Wrote a description of my changes above 
- [x] Formatted my code with [`black`](https://black.readthedocs.io/en/stable/index.html)
- [x] Added a bullet point for my changes to the top of the `CHANGELOG.md` file
- [x] Added or modified unit tests to reflect my changes
- [x] Manually tested with a notebook
- [x] If adding a feature, there is an example notebook and/or documentation in the `README.md` file

### CODE REVIEW

**Comment by venu1412** [comment]
@devstein Can you please review this?

File: sparkmagic/sparkmagic/livyclientlib/reliablehttpclient.py [hunk 1]
```diff
@@ -24,6 +25,20 @@ def __init__(self, endpoint, headers, retry_policy):
...
+        self._set_http_session_config()
+
+    def _set_http_session_config(self):
+        http_session_config = conf.http_session_config()
+        if http_session_config and http_session_config["adapters"]:
+            self._set_http_session_adapters(http_session_config["adapters"])
+
+    def _set_http_session_adapters(self, adapters):
+        for adapter in adapters:
+            full_class = adapter["adapter"]
```
Comment (devstein) [inline sparkmagic/sparkmagic/livyclientlib/reliablehttpclient.py:37]: Raise error if `adapter` key is missing?

**Review by devstein — COMMENTED** [review]
@venu1412 Great feature, thank you for contributing! 

Minor code question/changes for invalid adapter configs. Can you add some documentation for how to use this feature and some use-cases? The best place would either be in a new section in the top-level README or in an example notebook.

File: sparkmagic/sparkmagic/livyclientlib/reliablehttpclient.py [hunk 2]
```diff
@@ -24,6 +25,20 @@ def __init__(self, endpoint, headers, retry_policy):
                 "ATTENTION: Will ignore SSL errors. This might render you vulnerable to attacks."
             )
             requests.packages.urllib3.disable_warnings()
+        self._set_http_session_config()
+
+    def _set_http_session_config(self):
+        http_session_config = conf.http_session_config()
+        if http_session_config and http_session_config["adapters"]:
```
[the remark this answers is not in the document]
Comment (venu1412) [reply sparkmagic/sparkmagic/livyclientlib/reliablehttpclient.py]: Done

**Comment by venu1412** [comment]
@devstein i have resolved the comments and also added the the documentation on how to use it and when to use it. Can you please take a look again?

File: README.md [hunk 3]
```diff
@@ -156,6 +156,28 @@ After installing, you need to register the custom authenticator with Sparkmagic
               }
       ```
 
+## HTTP Session adapters
+
+when you need to customize HTTP request behavior for specific domains by modifying headers, implementing custom logic (e.g., using mTLS, retrying requests), or handling them differently. Utilize a custom adapter to gain fine-grained control over request processing.
+
+More details on how we can configure and use http adapter can be found [here](https://requests.readthedocs.io/en/latest/user/advanced/#transport-adapters)
```
Comment (devstein) [inline README.md:163]:
```suggestion
## HTTP Session Adapters

If you need to customize HTTP request behavior for specific domains by modifying headers, implementing custom logic (e.g., using mTLS, retrying requests), or handling them differently, you can use a custom adapter to gain fine-grained control over request processing.

More details on how we can configure and use http adapter can be found [here](https://requests.readthedocs.io/en/latest/user/advanced/#transport-adapters)
```

**Pull request merged**